File: ecmascript/binary/binary.js
export default class Binary {
    constructor(binary) {
        this.digits = /^[01]*$/.test(binary) ? [...binary] : [];
    }

    toDecimal() {
        return this.digits.reduce((acc, c) => acc * 2 + this.charToDecimal(c), 0)
    }

    charToDecimal(c) {
        return c - '0';
    }
}

File: prolog/binary/binary.pl
binary(Str, Dec) :- string_chars(Str, Digits), foldl(binary, Digits, 0, Dec).

binary('0', Acc, NewAcc) :- NewAcc is Acc << 1, !.
binary('1', Acc, NewAcc) :- NewAcc is Acc << 1 + 1.
